src/fomo.jl

```diff
@@ -0,0 +1,365 @@
+export fomo, FomoSolver, R2, R2Solver, tr, r2, R2og
+
+abstract type AbstractFomoMethod end
```

Comment:
What about we do the dispatch on the Solver type instead of creating fake structures?

It would work like this:
```suggestion
abstract type AbstractFirstOrderSolver <: AbstractOptimizationSolver end
```
and then later define
```
mutable struct FomoSolver{T, V} <: AbstractFirstOrderSolver
```
and same for R2.
In the `solve!` function the first argument would be `AbstractFirstOrderSolver` instead of `FomoSolver`, which makes more sense I think and helps adding new variants.
Later, you can replace `init_alpha(norm_∇fk::T, ::Union{r2,R2og})` by `init_alpha(solver, norm_∇fk::T)` to do the dispatch on the solver type.

What do you think @d-monnet ?

Maybe it doesn't perfectly fit with R2og, but I am not sure to fully understand the difference between R2 and R2og

Reply:
`R2og` is the original R2 algorithm. I use this backend to avoid unecessary computation in `solve!`. `r2` backend runs `solve!` as R2 with momentum, and `tr` backend runs `solve!` with momentum and first-order TR step.
They are 2 configurations to consider here:
1. Is momentum used (independent of R2 or TR) ? if yes, it is necessary to allocate extra vectors, which is currently done in `FomoSolver`. If no, for now `R2Solver()` returns a `FomoSolver` with no space allocated for the unecessary vectors.
2. Is R2 or TR step used ? This changes what `init_alpha` and `step_mult` should return.

The solver structure, which is meant to allocate memory space, is the same for TR and R2 steps, but differs whether momentum is used or not.
If we want to use dispatch for `init_alpha` and `step_mult`, which differs from TR and R2 but not whether momentum is used or not, it would be necessary to define lets say `R2MomentumSolver` and `TRMomentumSolver` which would be the exact same structure.

I think implementing a `AbstractFirstOrder` solver is a good idea, but I would define only two child structures `FOMOSolver` for momentum structure and `FOSolver` for no momentum structure, and keep a backend for TR or R2 step.

Reply:
I gave it more thoughts: If we define two solver structures for momentum and no-momentum, that will complexify `solve!` by adding more cases in the function.
For example, `-d` is the search direction which includes the momentum. If R2 (no momentum) is used, `R2Solver` defines `d = g` and it's transparent in `solve!`. If we define a `FOSolver` without `d` vector to store momentum, we'll have to modify `solve!` to handle this case.
I see two solutions:
1. either we leave things how they are with only one solver structure
2. we define 2 solver structures for momentum and no momentum, and we put the momentum-sensitive parts of `solve!` in dedicated functions with dispatch.

2 will create a bit more code, and `solve!` will be less straightforward to read, but it separates momentum and no-momentum more clearly. I'm ok with both solutions, I don't really have any preference.
@tmigot and @dpo let me know what you think.

Reply:
In the end I went for option 2, it doesn't really change `solve!` and makes things clearer in my opinion. 
`FoSolver` does not have docstring however, not sure what to put there.